File: app/src/main/java/im/point/dotty/domain/MainViewModel.java
package im.point.dotty.domain;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.ViewModelProvider;

import java.util.List;

import im.point.dotty.model.AllPost;
import im.point.dotty.model.CommentedPost;
import im.point.dotty.model.RecentPost;
import im.point.dotty.repository.RepoFactory;
import im.point.dotty.repository.Repository;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;

public final class MainViewModel extends AndroidViewModel {
    private RepoFactory repoFactory;
    private Repository<RecentPost> recentPostRepository;
    private Repository<CommentedPost> commentedPostRepository;
    private Repository<AllPost> allPostRepository;

    public MainViewModel(@NonNull Application application) {
        super(application);
        repoFactory = new RepoFactory(application.getBaseContext());
        recentPostRepository = repoFactory.getRecentRepo();
        commentedPostRepository = repoFactory.getCommentedRepo();
        allPostRepository = repoFactory.getAllRepo();
    }

    public Completable fetchRecent() {
        return Completable.fromSingle(recentPostRepository.fetch()).observeOn(AndroidSchedulers.mainThread());
    }

    public Flowable<List<RecentPost>> getRecent() {
        return recentPostRepository.getAll().observeOn(AndroidSchedulers.mainThread());
    }

    public Completable fetchAll() {
        return Completable.fromSingle(allPostRepository.fetch()).observeOn(AndroidSchedulers.mainThread());
    }

    public Flowable<List<AllPost>> getAll() {
        return allPostRepository.getAll().observeOn(AndroidSchedulers.mainThread());
    }

    public Completable fetchCommented() {
        return Completable.fromSingle(commentedPostRepository.fetch()).observeOn(AndroidSchedulers.mainThread());
    }

    public Flowable<List<CommentedPost>> getCommented() {
        return commentedPostRepository.getAll().observeOn(AndroidSchedulers.mainThread());
    }

}

File: app/src/main/java/im/point/dotty/domain/MainViewModel.kt
package im.point.dotty.domain

import androidx.lifecycle.AndroidViewModel
import im.point.dotty.DottyApplication
import im.point.dotty.model.AllPost
import im.point.dotty.model.CommentedPost
import im.point.dotty.model.RecentPost
import im.point.dotty.model.User
import im.point.dotty.repository.*
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers

class MainViewModel internal constructor(application: DottyApplication) : AndroidViewModel(application) {
    private val repoFactory: RepoFactory
    private val recentRepo: RecentRepo
    private val commentedPostRepo: CommentedRepo
    private val allPostRepo: AllRepo
    private val userRepo: UserRepo
    private val shared: Shared = Shared(application.state, application.mainApi)
    private val state: AppState = application.state

    fun fetchRecent(isBefore: Boolean): Completable {
        return Completable.fromSingle(if (isBefore) recentRepo.fetchBefore() else recentRepo.fetch()
                .flatMap { shared.fetchUnreadCounters() })
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun getRecent(): Flowable<List<RecentPost>> {
        return recentRepo.getAll().observeOn(AndroidSchedulers.mainThread())
    }

    fun fetchAll(isBefore: Boolean): Completable {
        return Completable.fromSingle(if (isBefore) allPostRepo.fetchBefore() else allPostRepo.fetch())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun getAll(): Flowable<List<AllPost>> {
        return allPostRepo.getAll().observeOn(AndroidSchedulers.mainThread())
    }

    fun fetchCommented(isBefore: Boolean): Completable {
        return Completable.fromSingle(if (isBefore) commentedPostRepo.fetchBefore() else commentedPostRepo.fetch())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun getCommented(): Flowable<List<CommentedPost>> {
        return commentedPostRepo.getAll().observeOn(AndroidSchedulers.mainThread())
    }

    fun getMe(): Flowable<User> {
        return userRepo.getMe().observeOn(AndroidSchedulers.mainThread())
    }

    fun getUser(id: Long): Flowable<User> {
        return userRepo.getItem(id).observeOn(AndroidSchedulers.mainThread())
    }

    fun fetchUser(id: Long): Single<User> {
        return userRepo.fetchUser(id).observeOn(AndroidSchedulers.mainThread())
    }

    fun fetchUnreadCounters(): Completable {
        return Completable.fromSingle(shared.fetchUnreadCounters())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun getUnreadPosts() = state.unreadPosts

    fun getUnreadComments() = state.unreadComments

    fun getUnreadPrivatePosts() = state.privateUnreadPosts

    fun getUnreadPrivateComments() = state.privateUnreadComments

    init {
        repoFactory = application.repoFactory
        recentRepo = repoFactory.getRecentRepo()
        commentedPostRepo = repoFactory.getCommentedRepo()
        allPostRepo = repoFactory.getAllRepo()
        userRepo = repoFactory.getUserRepo()
    }
}